src/Microsoft.ML.Core/Utilities/PathUtils.cs

```diff
@@ -139,5 +141,63 @@ public static string CreateFolderIfNotExists(string folder)
 
             return null;
         }
+
+        /// <summary>
+        /// Make a full path realtive to a base path.
+        /// </summary>
+        /// <param name="basepath">The base path, assumed to be a directory.</param>
+        /// <param name="path">The full path.</param>
+        /// <returns>The relative path.</returns>
+        /// <exception cref="ArgumentException">If the paths are not relative.</exception>
+        public static string MakePathRelative(string basepath, string path)
+        {
+            Contracts.AssertNonEmpty(basepath);
+            Contracts.AssertNonEmpty(path);
+
+            Uri baseUri = new Uri(basepath);
+            Uri uri = new Uri(path);
+
+            if (baseUri.Scheme != uri.Scheme)
+            {
+                throw new ArgumentException("Paths cannot be made relative as they are of different schemas.");
```

Comment:
>ArgumentException [](start = 26, length = 17)

can se use host.except instead of throw?

Reply:
As a Util class, it's my opinion it should be unaware of host. It would then be incumbent on the caller (who is aware of host) to pass the exception on to host.except. Is there a strong reason to avoid thowing that I'm not aware of?

Reply:
pretty please

---
In reply to: [187689960](https://github.com/dotnet/machinelearning/pull/61#discussion_r187689960) [](ancestors = 187689960)

Reply:
Hi Tyler! While we generally prefer that exceptions be thrown through host so that the context information of the call is preserved, we have not done this with the utils classes for one reason or another.

However, it is *still* important that the exception be marked. E.g.: use `Contracts.ExceptParam` here. This is how, say, the command line or GUI distinguishes between an error we are raising (which we don't necessarily consider a bug, but potentially a sign of misused) vs. unmarked exceptions. The latter, we *always* consider bugs.

---
In reply to: [187710640](https://github.com/dotnet/machinelearning/pull/61#discussion_r187710640) [](ancestors = 187710640)

Reply:
Note that if you're using `Contracts` instead, it would be something like:

```csharp
Contracts.CheckParam(baseUri.Scheme == uri.Scheme, nameof(basepath), "Paths cannot be made relative as they are of different schemes.");
```

Similar for the below, maybe.

---
In reply to: [189994300](https://github.com/dotnet/machinelearning/pull/61#discussion_r189994300) [](ancestors = 189994300,187710640)